agent/src/jvmMain/kotlin/com/epam/drill/agent/transport/RetryingTransportStateNotifier.kt

```diff
@@ -0,0 +1,62 @@
+/**
+ * Copyright 2020 - 2022 EPAM Systems
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.epam.drill.agent.transport
+
+import kotlin.concurrent.thread
+import mu.KotlinLogging
+
+class RetryingTransportStateNotifier<T>(
+    private val transport: AgentMessageTransport<T>,
+    private val messageSerializer: AgentMessageSerializer<T>,
+    private val messageQueue: AgentMessageQueue<T>
+) : TransportStateNotifier, TransportStateListener {
+
+    private val logger = KotlinLogging.logger {}
+    private val stateListeners = mutableSetOf<TransportStateListener>()
+    internal var retrying = false
+    internal var retryingThread: Thread? = null
+
+    override fun addStateListener(listener: TransportStateListener) {
+        stateListeners.add(listener)
+    }
+
+    override fun onStateAlive() {
+        logger.debug { "onStateAlive: Alive event received" }
+    }
+
+    override fun onStateFailed() = synchronized(this) {
+        logger.debug { "onStateFailed: Failed event received, current retrying state: $retrying" }
+        if (!retrying) {
+            messageQueue.element()

```

Comment:
`element()` return value is unused - it's unclear _why_.
I guess the intent is to generate exception in case of empty queue to interrupt code execution.
If that's the case - better do it without exceptions.
I suggest to check it in some other way.
```
if (messageQueue.element().isNull()) return
// or just check for queue size directly
if (messageQueue.size() == 0) return 
```

Reply:
In this case it should be un-catched exceptions, because if it was thrown - the code is wrong.
It's some kind of indicator: if we call failed message retrying process, code expect that this failed message exists in queue, if not - something coded wrong.
But I agree, this is unclear :)